diff --git a/dist/index.js b/dist/index.js
index 57f53f1..5e1c7f4 100644
--- a/dist/index.js
+++ b/dist/index.js
@@ -78,12 +78,8 @@ function runFixtureTest(name, dir, run, mode, context) {
     const skip = main.skip || main["skip_" + mode];
     const fails = main.fails || main["fails_" + mode];
 
-    if (skip) {
-      mochaTestFunction = it.skip;
-      mochaDetails = skip;
-    } else if (fails) {
-      mochaTestFunction = it.fails;
-      mochaDetails = fails;
+    if (skip || fails) {
+      return;
     }
   }
 
@@ -128,10 +124,7 @@ function runFixtureTest(name, dir, run, mode, context) {
   };
 
   const skip = reason => {
-    const test = it.skip(name);
-    test.details = reason;
-    test.file = mainPath;
-    return test;
+    return { details: reason, file: mainPath };
   };
 
   run({